crates/ika-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -620,4 +652,67 @@ impl DWalletMPCManager {
             _ => 0,
         }
     }
+
+    /// Builds the outputs to finalize based on the outputs received in the consensus rounds.
+    /// If a majority vote is reached, it returns the malicious voters (didn't vote with majority) and the majority vote.
+    /// If the threshold is not reached, it returns `None`.
+    pub(crate) fn build_outputs_to_finalize(
+        &self,
+        session_identifier: &SessionIdentifier,
+        outputs_by_consensus_round: HashMap<
+            u64,
+            HashMap<PartyID, Vec<DWalletCheckpointMessageKind>>,
+        >,
+    ) -> Option<(Vec<PartyID>, Vec<DWalletCheckpointMessageKind>)> {
+        let mut outputs_to_finalize: HashMap<PartyID, Vec<DWalletCheckpointMessageKind>> =
+            HashMap::new();
+
+        for (_, outputs) in outputs_by_consensus_round {
+            for (sender_party_id, output) in outputs {
+                // take the last output from each sender party ID
+                outputs_to_finalize.insert(sender_party_id, output);
+            }
+        }
+
+        match outputs_to_finalize.weighted_majority_vote(&self.access_structure) {
+            Ok((malicious_voters, majority_vote)) => Some((malicious_voters, majority_vote)),
```

Comment:
lop the consensus round, the malicious voters, the majority vote, and total voters